crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -377,48 +382,45 @@ where
 
         let topics = filter.has_topics().then(|| filter_params.flat_topics.clone());
 
-        // derive bloom filters from filter input
-        let address_filter = FilteredParams::address_filter(&filter.address);
-        let topics_filter = FilteredParams::topics_filter(&topics);
+        // Create log index filter
+        let mut log_index_filter =
+            LogIndexFilter::new(from_block..=to_block, self.max_headers_range);
+        log_index_filter.install_address_filter(&self.provider, filter.address.as_ref())?;
+        log_index_filter.install_topic_filter(&self.provider, topics.as_ref())?;
 
         // loop over the range of new blocks and check logs if the filter matches the log's bloom
         // filter
-        for (from, to) in
-            BlockRangeInclusiveIter::new(from_block..=to_block, self.max_headers_range)
-        {
-            let headers = self.provider.headers_range(from..=to)?;
+        for block_numbers in log_index_filter.iter() {
+            let headers = self.provider.headers(&block_numbers)?;
 
             for (idx, header) in headers.iter().enumerate() {
+                let header = match header {
+                    Some(header) => header,
+                    None => continue,
+                };
+
                 // these are consecutive headers, so we can use the parent hash of the next block to
                 // get the current header's hash
                 let num_hash: BlockHashOrNumber = headers
                     .get(idx + 1)
-                    .map(|h| h.parent_hash.into())
+                    .map(|h| h.as_ref().unwrap().parent_hash.into())
```

Comment:
Let's add a note on safety of the unwrap due to the continue above

Reply:
looked over this again, it's not actually safe and is sometimes incorrect. fixing